File: sdk/python-can-gsusb/examples/01_simple_std.py
import can

from common import get_parser
from gsusb import GsUsbBus


def main() -> None:
    parser = get_parser("Single-channel classic CAN example")
    parser.add_argument("--gsusb-channel", type=int, default=0, help="Logical gs_usb channel (default: 0)")
    args = parser.parse_args()

    print(f"Opening {args.selector} Channel {args.gsusb_channel} @ {args.bitrate}bps...")
    bus = GsUsbBus(
        channel=args.selector,
        gsusb_channel=args.gsusb_channel,
        bitrate=args.bitrate,
        fd=False,
        termination_enabled=not args.no_termination,
    )
    try:
        msg = can.Message(arbitration_id=0x123, data=[0x11, 0x22, 0x33, 0x44], is_extended_id=False)
        print(f"Sending: {msg}")
        bus.send(msg)
        print("Listening for messages (Ctrl+C to exit)...")
        while True:
            rx = bus.recv(1.0)
            if rx is not None:
                print(f"Received: {rx}")
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        bus.shutdown()


if __name__ == "__main__":
    main()

File: sdk/gsusb-rs/examples/01_simple_std.rs
#[path = "common/mod.rs"]
mod common;

use common::{make_bus_options, parse_common_options, print_message};
use gsusb_rs::{CanMessage, GsUsbBus, Result};
use std::env;
use std::time::Duration;

fn main() -> Result<()> {
    let args = env::args().skip(1).collect::<Vec<_>>();
    let options = parse_common_options(&args, "Single-channel classic CAN example")?;
    println!("Opening {} tx={} rx={} @ {}bps", options.selector, options.tx_channel, options.rx_channel, options.bitrate);
    let tx_bus = GsUsbBus::new(make_bus_options(&options, options.tx_channel, false, None, None))?;
    let rx_bus = GsUsbBus::new(make_bus_options(&options, options.rx_channel, false, None, None))?;

    let message = CanMessage {
        arbitration_id: 0x123,
        data: vec![0x11, 0x22, 0x33, 0x44],
        ..CanMessage::default()
    };
    print_message("Sending", &message);
    tx_bus.send(&message)?;
    let received = rx_bus.recv(Duration::from_secs(2)).ok_or_else(|| gsusb_rs::GsUsbError("No frame received".to_string()))?;
    print_message("Received", &received);
    tx_bus.shutdown();
    rx_bus.shutdown();
    Ok(())
}